<#30>feat: implement page to allow for editing of villages

khkhiu opened the pull request:
**Description:**
This PR add a 'settings' page to allow admins to view and edit available villages. Do note that villages cannot be deleted to maintain compatibility with past data, but can be hidden instead.

**Features:**
- Create DynamicVillageDropdown component for village selection
- Implement soft delete functionality (hide/show villages)
- Add Settings page with village administration interface
- Update VillageContext to support dynamic village data
- Integrate village management with existing search components

**Issue ticket number and link:**
https://github.com/NUS-Project-SaBai/FrontEnd/issues/30

khkhiu committed:
feat(villages): implement village management system
- Create DynamicVillageDropdown component for village selection
- Implement soft delete functionality (hide/show villages)
- Add Settings page with village administration interface
- Update VillageContext to support dynamic village data
- Integrate village management with existing search components

pohanson commented on src/hooks/useVillages.ts:
```diff
@@ -0,0 +1,46 @@
+  village_name: string;
+  colour_code: string;
+  is_hidden: boolean;
+}
```
Would there be any reason for redefining the `Village` instead of importing from the `VillageTypes`?

pohanson commented on src/app/settings/page.tsx:
The UI of the pages is very nice! ✨ 

Although there are comments to help annotate and make it readable, this file is too long and complex, making it harder to manage. Would it be possible to separate this page into components? Perhaps it could be separated into the form component, and the table component. Then for the table component, it could possibly be further separated into rows if the component is too complex.

pohanson commented on src/app/settings/page.tsx:
```diff
@@ -0,0 +1,543 @@
+
+        <div className="p-6">
+          {/* Add Village Form - Only shown when showAddForm is true */}
+          {showAddForm && (
```
Would it be possible to use React Hook Form? We have a few RHF component in `src/components/inputs` so that we can have a consistent look. Another benefit is that RHF is able to handle validation more easily. 

View `AddMedicationModal.tsx` and `MedicationForm.tsx`, or any other forms for an example usage.

pohanson commented on src/components/PatientSearchbar.tsx:
```diff
@@ -38,7 +38,7 @@ export function PatientSearchbar<T>({
       setFilteredItems(data);
     }
-  }, [queryStr, setFilteredItems, village, data]);
+  }, [queryStr, setFilteredItems, village, data, filterFunction]);
```
The filterFunction seems to be created on each render, resulting in useEffect constantly being ran.
```suggestion
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [queryStr, setFilteredItems, village, data]);
```

pohanson commented on src/components/referrals/PatientSearchbarByReferral.tsx:
```diff
@@ -44,9 +44,10 @@ export function PatientSearchInputByReferral({
 
-    if (village != VillagePrefix.ALL) {
+    // Filter by village if not "ALL"
+    if (village !== 'ALL') {
```
Perhaps a constant defined in `VillageTypes` could be used?

pohanson commented on src/context/VillageContext.tsx:
```diff
@@ -1,22 +1,28 @@
-  village: VillagePrefix.ALL,
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  setVillage: (village: VillagePrefix) => {},
+  village: 'ALL' as string,
```
Perhaps you could use the constant you had defined in `VillageTypes` to replace `'ALL'`?

pohanson commented on src/app/settings/page.tsx:
```diff
@@ -0,0 +1,543 @@
+import { AxiosError } from 'axios';
+import { useEffect, useState } from 'react';
+
+interface Village {
+  id: number;
+  village_name: string;
+  colour_code: string;
+  is_hidden: boolean;
+}
```
Would there be any reasons that you've redefined `Village` instead of importing?

pohanson reviewed (changes requested):
Good job! 👍 It looks nice, and code is readable. 

p.s. This feature might not be merged into for this year trip because of the difficulty of ensuring the data is properly migrated with the village table.

khkhiu committed:
Refactor: Use shared Village type and ALL_VILLAGES constant:

- Remove duplicate Village interface definitions
- Import Village type from VillageTypes
- Replace hardcoded 'ALL' strings with ALL_VILLAGES constant

khkhiu committed:
Refactor(settings): separate village management into reusable components

- Split monolithic settings page into AddVillageForm, VillageTable, and VillageTableRow components
- Integrate React Hook Form with existing RHF input components for consistent validation

khkhiu committed:
fix: memoize filterFunction in PatientSearchbar parent components

Wrap filterFunction with useCallback in Orders, Records, and Vision pages
to prevent infinite re-render loop. The inline function was being recreated
on every render, causing PatientSearchbar's useEffect to trigger repeatedly.
